File: 백준/Gold/2056. 작업/작업.cc
#include <iostream>
#include <vector>
#include <queue>

#define PII std::pair<int, int>

int N;
int costs[10001];
int finish_costs[10001];
int in_degree[10001];
std::vector<int> out_degree[10001];

int main() {
    std::cin >> N;

    int cost, pre_works, pre_work;
    for (int i = 1; i < N + 1; i++) {
        std::cin >> cost >> pre_works;
        costs[i] = cost;
        in_degree[i] = pre_works;
        for (int j = 0; j < pre_works; j++) {
            std::cin >> pre_work;
            out_degree[pre_work].push_back(i);
        }
    }

    std::queue<int> q;
    for (int i = 1; i < N + 1; i++) {
        if (in_degree[i] == 0) {
            q.push(i);
            finish_costs[i] = costs[i];
        }
    }

    while (!q.empty()) {
        int cur = q.front();
        q.pop();

        for (int next : out_degree[cur]) {
            finish_costs[next] = std::max(finish_costs[next], finish_costs[cur] + costs[next]);
            in_degree[next]--;
            if (in_degree[next] == 0) {
                q.push(next);
            }
        }
    }

    int answer = 0;
    for (int i = 1; i < N + 1; i++) {
        answer = std::max(answer, finish_costs[i]);
    }
    std::cout << answer << "\n";

    return 0;
}

File: 백준/Gold/2056. 작업/작업.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static int N;
    static int[] times;
    static int[] pre;
    static List<Integer>[] post;
    static int[] finishTimes;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        N = Integer.parseInt(br.readLine());
        times = new int[N + 1];
        pre = new int[N + 1];
        post = new ArrayList[N + 1];
        for (int i = 1; i < N + 1; i++) {
            post[i] = new ArrayList<>();
        }
        finishTimes = new int[N + 1];

        StringTokenizer stk;
        int preWorks = 0, preWork = 0;
        for (int i = 1; i < N + 1; i++) {
            stk = new StringTokenizer(br.readLine());
            times[i] = Integer.parseInt(stk.nextToken());
            preWorks = Integer.parseInt(stk.nextToken());
            pre[i] = preWorks;
            for (int j = 0; j < preWorks; j++) {
                preWork = Integer.parseInt(stk.nextToken());
                post[preWork].add(i);
            }
        }

        Queue<Integer> q = new ArrayDeque<>();
        for (int i = 1; i < N + 1; i++) {
            if (pre[i] == 0) {
                q.add(i);
                finishTimes[i] = times[i];
            }
        }

        while (!q.isEmpty()) {
            int cur = q.peek();
            q.poll();

            for (int i = 0; i < post[cur].size(); i++) {
                int next = post[cur].get(i);
                finishTimes[next] = Math.max(finishTimes[next], finishTimes[cur] + times[next]);
                pre[next]--;
                if (pre[next] == 0) {
                    q.add(next);
                }
            }
        }

        int answer = 0;
        for (int i = 1; i < N + 1; i++) {
            answer = Math.max(answer, finishTimes[i]);
        }

        System.out.println(answer);
    }
}